### PR TITLE
add deviation detection

### PR DESCRIPTION
**Pull request by kklgf**

### CODE REVIEW

File: spring-app/src/main/java/pl/edu/agh/cqm/service/impl/SchedulerService.java [hunk 1]
```diff
@@ -18,12 +18,12 @@
...
+//    @Scheduled(fixedDelay = 1)
     public void scheduledThroughput(){
         throughputService.doMeasurement();
     }
 
-    @Scheduled(fixedRate = 60000)
+//    @Scheduled(fixedRate = 60000)
     public void schedulePing() {
         if (schedulePingCounter == 0) {
             pingService.doMeasurement();
```
Comment (HubertBalcerzak) [inline spring-app/src/main/java/pl/edu/agh/cqm/service/impl/SchedulerService.java]: I think this wasn't supposed to be commited

File: spring-app/src/main/java/pl/edu/agh/cqm/service/impl/DeviationsServiceImpl.java [hunk 2]
```diff
@@ -94,12 +94,78 @@
...
+                        slowChanges.add(new DeviationDTO(values.get(is).getFirst(), values.get(ie).getFirst(), "Increase"));
+                    }
+                }
+            }
+            return slowChanges;
+        }
+        return List.of();
+    }
+
+    private List<DeviationDTO> rappidChanges(List<Pair<Instant, Number>> values, double rappidChangePercentage) {
```
Comment (HubertBalcerzak) [inline spring-app/src/main/java/pl/edu/agh/cqm/service/impl/DeviationsServiceImpl.java]: rapid

File: spring-app/src/main/java/pl/edu/agh/cqm/service/impl/DeviationsServiceImpl.java [hunk 3]
```diff
@@ -94,12 +94,78 @@
...
+        if (values.size() > 1) {
+            List<DeviationDTO> rappidChanges = new ArrayList<>();
```
Comment (HubertBalcerzak) [inline spring-app/src/main/java/pl/edu/agh/cqm/service/impl/DeviationsServiceImpl.java]: rapid

File: spring-app/src/main/java/pl/edu/agh/cqm/service/impl/DeviationsServiceImpl.java [hunk 4]
```diff
@@ -94,12 +94,78 @@
...
+                        slowChanges.add(new DeviationDTO(values.get(is).getFirst(), values.get(ie).getFirst(), "Decrease"));
+                    }
+                } else if (values.get(i - 1).getSecond().doubleValue() < values.get(i).getSecond().doubleValue()) {
+                    int is = i - 1;
+                    while (i < values.size() && values.get(i - 1).getSecond().doubleValue() < values.get(i).getSecond().doubleValue()) {
+                        i++;
+                    }
+                    i--;
+                    int ie = i;
+                    if (~ i < values.size()){
```
Comment (HubertBalcerzak) [inline spring-app/src/main/java/pl/edu/agh/cqm/service/impl/DeviationsServiceImpl.java]: what is this condition even checking?

File: spring-app/src/main/java/pl/edu/agh/cqm/service/impl/DeviationsServiceImpl.java [hunk 5]
```diff
@@ -94,12 +94,78 @@
...
+            return deviations;
         }
     }
+
+    private List<DeviationDTO> slowChanges(List<Pair<Instant, Number>> values, double slowChangeMinLength) {
+        if (values.size() > slowChangeMinLength) {
+            List<DeviationDTO> slowChanges = new ArrayList<>();
+            for (int i = 1; i < values.size(); i++) {
+                if (values.get(i - 1).getSecond().doubleValue() > values.get(i).getSecond().doubleValue()) {
+                    int is = i - 1;
```
Comment (HubertBalcerzak) [inline spring-app/src/main/java/pl/edu/agh/cqm/service/impl/DeviationsServiceImpl.java]: variable names should be self documenting

File: spring-app/src/main/java/pl/edu/agh/cqm/service/impl/DeviationsServiceImpl.java [hunk 6]
```diff
@@ -94,12 +94,78 @@
...
     }
 
-    // TODO - CQM-51
     private List<DeviationDTO> detectDeviations(List<Pair<Instant, Number>> values) {
         if (values.isEmpty()) {
             return List.of();
         } else {
-            return List.of(new DeviationDTO(values.get(0).getFirst(), values.get(0).getFirst().plusSeconds(1), "test"));
+            List<DeviationDTO> deviations = new ArrayList<>(rappidChanges(values, 0.2));
+            deviations.addAll(slowChanges(values, 5));
```
Comment (HubertBalcerzak) [inline spring-app/src/main/java/pl/edu/agh/cqm/service/impl/DeviationsServiceImpl.java]: put magic numbers in static constants

File: spring-app/src/main/java/pl/edu/agh/cqm/service/impl/DeviationsServiceImpl.java [hunk 7]
```diff
@@ -94,12 +94,78 @@
...
+        if (values.size() > slowChangeMinLength) {
+            List<DeviationDTO> slowChanges = new ArrayList<>();
+            for (int i = 1; i < values.size(); i++) {
+                if (values.get(i - 1).getSecond().doubleValue() > values.get(i).getSecond().doubleValue()) {
+                    int is = i - 1;
+                    while (i < values.size() && values.get(i - 1).getSecond().doubleValue() > values.get(i).getSecond().doubleValue()) {
+                        i++;
+                    }
+                    i--;
+                    int ie = i;
```
Comment (HubertBalcerzak) [inline spring-app/src/main/java/pl/edu/agh/cqm/service/impl/DeviationsServiceImpl.java]: variable names should be self documenting

File: spring-app/src/main/java/pl/edu/agh/cqm/service/impl/DeviationsServiceImpl.java [hunk 8]
```diff
@@ -94,12 +94,78 @@
...
+                    if (~ i < values.size()){
+                        ie--;
+                    }
+                    if (ie - is >= slowChangeMinLength) {
+                        slowChanges.add(new DeviationDTO(values.get(is).getFirst(), values.get(ie).getFirst(), "Decrease"));
+                    }
+                } else if (values.get(i - 1).getSecond().doubleValue() < values.get(i).getSecond().doubleValue()) {
+                    int is = i - 1;
```
Comment (HubertBalcerzak) [inline spring-app/src/main/java/pl/edu/agh/cqm/service/impl/DeviationsServiceImpl.java]: variable names should be self documenting

File: spring-app/src/main/java/pl/edu/agh/cqm/service/impl/DeviationsServiceImpl.java [hunk 9]
```diff
@@ -94,12 +94,78 @@
...
+                                (1 - rappidChangePercentage) * values.get(i).getSecond().doubleValue()) {
+                    int is = i - 1;
+                    while (i < values.size() && (values.get(i - 1).getSecond().doubleValue() >
+                            (1 + rappidChangePercentage) * values.get(i).getSecond().doubleValue() ||
+                            values.get(i - 1).getSecond().doubleValue() <
+                                    (1 - rappidChangePercentage) * values.get(i).getSecond().doubleValue())) {
+                        i++;
+                    }
+                    i--;
+                    int ie = i;
```
Comment (HubertBalcerzak) [inline spring-app/src/main/java/pl/edu/agh/cqm/service/impl/DeviationsServiceImpl.java]: variable names should be self documenting

File: spring-app/src/main/java/pl/edu/agh/cqm/service/impl/DeviationsServiceImpl.java [hunk 10]
```diff
@@ -94,12 +94,78 @@
...
+
+    private List<DeviationDTO> rappidChanges(List<Pair<Instant, Number>> values, double rappidChangePercentage) {
+        if (values.size() > 1) {
+            List<DeviationDTO> rappidChanges = new ArrayList<>();
+            for (int i = 1; i < values.size(); i++) {
+                if (values.get(i - 1).getSecond().doubleValue() >
+                        (1 + rappidChangePercentage) * values.get(i).getSecond().doubleValue() ||
+                        values.get(i - 1).getSecond().doubleValue() <
+                                (1 - rappidChangePercentage) * values.get(i).getSecond().doubleValue()) {
+                    int is = i - 1;
```
Comment (HubertBalcerzak) [inline spring-app/src/main/java/pl/edu/agh/cqm/service/impl/DeviationsServiceImpl.java]: variable names should be self documenting

File: spring-app/src/main/java/pl/edu/agh/cqm/service/impl/DeviationsServiceImpl.java [hunk 11]
```diff
@@ -94,12 +94,78 @@
...
+            deviations.addAll(slowChanges(values, 5));
+            return deviations;
         }
     }
+
+    private List<DeviationDTO> slowChanges(List<Pair<Instant, Number>> values, double slowChangeMinLength) {
+        if (values.size() > slowChangeMinLength) {
+            List<DeviationDTO> slowChanges = new ArrayList<>();
+            for (int i = 1; i < values.size(); i++) {
+                if (values.get(i - 1).getSecond().doubleValue() > values.get(i).getSecond().doubleValue()) {
```
Comment (HubertBalcerzak) [inline spring-app/src/main/java/pl/edu/agh/cqm/service/impl/DeviationsServiceImpl.java]: there are a lot of these `.getSecond().doubleValue()` expressions here. Maybe introduce a helper method?

File: spring-app/src/main/java/pl/edu/agh/cqm/service/impl/DeviationsServiceImpl.java [hunk 12]
```diff
@@ -94,12 +94,78 @@
...
+    }
+
+    private List<DeviationDTO> rappidChanges(List<Pair<Instant, Number>> values, double rappidChangePercentage) {
+        if (values.size() > 1) {
+            List<DeviationDTO> rappidChanges = new ArrayList<>();
+            for (int i = 1; i < values.size(); i++) {
+                if (values.get(i - 1).getSecond().doubleValue() >
+                        (1 + rappidChangePercentage) * values.get(i).getSecond().doubleValue() ||
+                        values.get(i - 1).getSecond().doubleValue() <
+                                (1 - rappidChangePercentage) * values.get(i).getSecond().doubleValue()) {
```
Comment (HubertBalcerzak) [inline spring-app/src/main/java/pl/edu/agh/cqm/service/impl/DeviationsServiceImpl.java]: I'd extract such complex conditions to methods. That would give them names and improve readability

File: spring-app/src/main/java/pl/edu/agh/cqm/service/impl/DeviationsServiceImpl.java [hunk 13]
```diff
@@ -94,12 +94,78 @@
...
+                    }
+                }
+            }
+            return slowChanges;
+        }
+        return List.of();
+    }
+
+    private List<DeviationDTO> rappidChanges(List<Pair<Instant, Number>> values, double rappidChangePercentage) {
+        if (values.size() > 1) {
```
Comment (HubertBalcerzak) [inline spring-app/src/main/java/pl/edu/agh/cqm/service/impl/DeviationsServiceImpl.java:171]: !values.isEmpty()

**Review by HubertBalcerzak — CHANGES_REQUESTED** [review]

File: spring-app/src/main/java/pl/edu/agh/cqm/service/impl/DeviationsServiceImpl.java (hunk 13 above)
Comment (kklgf) [reply spring-app/src/main/java/pl/edu/agh/cqm/service/impl/DeviationsServiceImpl.java:171]: values.size() have to be greater than 1, it cannot be equal to 1

**Pull request merged**